### DISCUSSION TITLE
Using AddListMembersListRequest

### DISCUSSION THREAD

**Discussion by CobraChickin** [jordimontana82/fake-xrm-easy#559] [Q&A] [answered]
Hi. I have a codeactivity that leverages `AddListMembersListRequest()` (or `AddMemberListRequest()`) and works within CRM 2013. When running unit tests, the code passes through the request and the response has empty Guid (when I use either request) - and does nothing. There is no "not supported" message - is this supported in FakeXrmEasy2013? If yes, I can attach screen of my mock.

**Comment by BetimBeja** [comment, marked as answer]
As you can see [here](https://github.com/jordimontana82/fake-xrm-easy/blob/f294b9af7319f8e23d4032a9290ec75a628f2435/FakeXrmEasy.Tests.Shared/FakeContextTests/AddListMembersListRequestTests/Tests.cs) there are tests that describe the functionality of `AddListMembersListRequest`.
If these tests don't help please write your test in a code block comment using ```csharp and I can try and help

**Reply by CobraChickin** [reply]
The CodeActivity calls `list.AddMember()`. This works in CRM and the Contact is added. I use XrmToolBox CrmEarlyBound generator for the entities.

```csharp
public static bool AddMember(this List list, CrmServiceContext service, Account account)
{
    if (list.HasMember(service, account))
        return false;   // return failure
            
    AddMemberListRequest req = new AddMemberListRequest();// new AddListMembersListRequest ???
    req.EntityId = account.PrimaryContactId.Id;
    req.ListId = list.Id;

    var resp = service.Execute(req) as AddMemberListResponse;
    return resp.Id != Guid.Empty;   // return success or failure
}

public static bool HasMember(this List list, CrmServiceContext service, Account account)
{
    return service.ListMemberSet
        .Where(x =>
            x.Id == account.PrimaryContactId.Id &&
            x.ListId.Id == list.Id)
        .ToList()
        .Any();
}
```
My mock consists of a SalesOrder, Account (customer on order), Contact (account's primary contact) etc.
```csharp
public void TestInitialize()
        {
            // simulate CRM
            context = new XrmFakedContext();
            wfContext = context.GetDefaultWorkflowContext();
            wfContext.MessageName = "Update";

            // Initialize Entities
            var entities = new List<Entity>();
            
            //... some other entities (products, salesorderdetails, etc)
            var fredContact = new Contact()
            {
                Id = Guid.NewGuid(),
                FirstName = "Fred",
                LastName = "Flinstone"
            };
            entities.Add(fredContact);
            var fred = new Account()
            {
                Id = Guid.NewGuid(),
                Name = "Fred",
                PrimaryContactId = new EntityReference(Contact.EntityLogicalName, fredContact.Id)
            };
            entities.Add(fred);
            var fredOrder = new SalesOrder()
            {
                Id = Guid.NewGuid(),
                Name = fred.Name + " Order",
                CustomerId = new EntityReference(Account.EntityLogicalName, fred.Id)
            };
            entities.Add(fredOrder);
            var list100 = new List()
            {
                Id = Guid.NewGuid(),
                ListName = "100 - blahblah",
                CreatedFromCode = List_CreatedFromCode.Contact,
                vse_wayofuse = List_vse_wayofuse.List,
                MemberType = (int)List_CreatedFromCode.Contact // dunno if needed
            };
            entities.Add(list100);
            context.Initialize(entities);
            
            // Default InputOutput Parameters
            wfContext.InputParameters = new ParameterCollection
            {
                { "RegardingOrder", new EntityReference(SalesOrder.EntityLogicalName, fredOrder.Id) }
            };
}
[TestMethod]
public void Pass_SingleOrder()
{
    var inputs = wfContext.InputParameters.ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
    context.ExecuteCodeActivity<ThankYouMessageMarketingListAssignment>(wfContext, inputs);

    var service = new CrmServiceContext(context.GetOrganizationService());

    Account fred = service.AccountSet.Where(x => x.Name == "Fred").FirstOrDefault();

    var list100 = service.ListSet.Where(x => x.ListName == "100 - blahblah").First();
    var list101 = service.ListSet.Where(x => x.ListName == "101 - blahblah").First();
    var list102 = service.ListSet.Where(x => x.ListName == "102 - blahblah").First();
    var list103 = service.ListSet.Where(x => x.ListName == "103 - blahblah").First();
    var list104 = service.ListSet.Where(x => x.ListName == "104 - blahblah").First();
    
    Assert.IsTrue(
        !list100.HasMember(service, fred) &&
        list101.HasMember(service, fred) && // should be added here, he is not
        !list102.HasMember(service, fred) &&
        list103.HasMember(service, fred) && // should be added here, he is not
        !list104.HasMember(service, fred));
}
```
Sorry for the wall of code, tried to give as much context as I could. I can see that the execution reached the request from debugging, but does nothing. Is there any Entities I need to mock? I tried mocking ListMember too, but didn't help. Thanks for help.

EDIT: I should also mention that I also tried using `AddListMembersListRequest`, with no success.

**Reply by BetimBeja** [reply]
I think the problem is with these lines of code, in you `AddMember` extension:
```csharp
   var resp = service.Execute(req) as AddMemberListResponse;
   return resp.Id != Guid.Empty;   // return success or failure
```

because the AddMemberListResponse returned by FakeXrmEasy does not specify the Id
https://github.com/jordimontana82/fake-xrm-easy/blob/f294b9af7319f8e23d4032a9290ec75a628f2435/FakeXrmEasy.Shared/FakeMessageExecutors/AddMemberListRequestExecutor.cs#L96

@jordimontana82  perhaps we should fix this 😄

**Reply by CobraChickin** [reply]
Hmm.. I haven't tested, whether it returned True on success. I've just seen False on failures. But that doesn't really solve the issue, that a Member isn't added to the MarketinList :/ .

**Reply by jordimontana82** [reply]
@Houkouha for the time being, and while v2 is not yet released, you could grab the fake message executor from this repo into your own test project, and override it with context.AddFakeMessageExecutor() with the necessary fix suggested above by @BetimBeja. We're not pushing any releases / fixes until v2 is out.

**Reply by jordimontana82** [reply]
It would be good a PR too if you manage to fix it there @Houkouha
